File: isabelle.eclipse.ui/src/isabelle/eclipse/ui/editors/TheoryEditor.java
package isabelle.eclipse.ui.editors;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.Collections;
import java.util.EnumSet;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

import isabelle.Command;
import isabelle.Document.Snapshot;
import isabelle.Session.Commands_Changed;
import isabelle.Session;
import isabelle.Thy_Header;
import isabelle.Thy_Info;
import isabelle.eclipse.core.IsabelleCorePlugin;
import isabelle.eclipse.core.app.IIsabelleSessionListener;
import isabelle.eclipse.core.app.Isabelle;
import isabelle.eclipse.core.resource.URIThyLoad;
import isabelle.eclipse.core.resource.URIPathEncoder;
import isabelle.eclipse.core.text.DocumentModel;
import isabelle.eclipse.core.text.DocumentModel$;
import isabelle.eclipse.core.util.SafeSessionActor;
import isabelle.eclipse.core.util.SessionEventSupport;
import isabelle.eclipse.ui.IsabelleUIPlugin;
import isabelle.eclipse.ui.text.DocumentListenerSupport;
import isabelle.eclipse.ui.views.TheoryOutlinePage;
import isabelle.scala.DocumentRef;
import isabelle.scala.ISessionCommandsListener;
import isabelle.scala.SessionActor;
import isabelle.scala.SessionEventType;
import isabelle.scala.TheoryInfoUtil;

import org.eclipse.core.filesystem.EFS;
import org.eclipse.core.filesystem.IFileStore;
import org.eclipse.core.filesystem.URIUtil;
import org.eclipse.core.resources.IFile;
import org.eclipse.core.resources.IResource;
import org.eclipse.core.resources.ResourcesPlugin;
import org.eclipse.core.runtime.Assert;
import org.eclipse.core.runtime.CoreException;
import org.eclipse.core.runtime.IPath;
import org.eclipse.core.runtime.IProgressMonitor;
import org.eclipse.core.runtime.IStatus;
import org.eclipse.core.runtime.Status;
import org.eclipse.jface.dialogs.MessageDialog;
import org.eclipse.jface.text.BadLocationException;
import org.eclipse.jface.text.DocumentEvent;
import org.eclipse.jface.text.IDocument;
import org.eclipse.jface.text.IDocumentListener;
import org.eclipse.jface.text.IRegion;
import org.eclipse.jface.text.ITextViewer;
import org.eclipse.jface.text.IViewportListener;
import org.eclipse.jface.text.JFaceTextUtil;
import org.eclipse.jface.text.Region;
import org.eclipse.jface.text.source.ILineRange;
import org.eclipse.jface.text.source.ISourceViewer;
import org.eclipse.swt.events.ControlAdapter;
import org.eclipse.swt.events.ControlEvent;
import org.eclipse.swt.events.ControlListener;
import org.eclipse.swt.widgets.Composite;
import org.eclipse.swt.widgets.Control;
import org.eclipse.ui.IEditorInput;
import org.eclipse.ui.IEditorPart;
import org.eclipse.ui.IEditorSite;
import org.eclipse.ui.PartInitException;
import org.eclipse.ui.contexts.IContextService;
import org.eclipse.ui.editors.text.ILocationProvider;
import org.eclipse.ui.editors.text.ILocationProviderExtension;
import org.eclipse.ui.editors.text.TextEditor;
import org.eclipse.ui.ide.ResourceUtil;
import org.eclipse.ui.progress.UIJob;
import org.eclipse.ui.texteditor.IDocumentProvider;
import org.eclipse.ui.views.contentoutline.IContentOutlinePage;

import scala.Option;
import scala.Tuple2;


public class TheoryEditor extends TextEditor {

	public static final String EDITOR_ID = "isabelle.eclipse.ui.theoryEditor";
	public static final String EDITOR_SCOPE = "isabelle.eclipse.ui.theoryEditorScope";
	
	private ColorManager colorManager;
	
	private final IIsabelleSessionListener sessionListener;
	
	private Session isabelleSession;
	
	private DocumentRef documentRef;
	private DocumentModel isabelleModel;
	
	/**
	 * A listener for scrolling events in the editor. Updates the active
	 * perspective upon scrolling.
	 */
	private final IViewportListener viewerViewportListener;
	
	/**
	 * A listener for resize events in the editor. Updates the active
	 * perspective upon editor resize.
	 */
	private final ControlListener viewerControlListener;
	
	/**
	 * A listener for document events in the editor. Updates the active
	 * perspective upon document modification.
	 */
	private DocumentListenerSupport documentListener;
	
	private TheoryOutlinePage outlinePage = null;
	
	private final TheoryAnnotations markers = new TheoryAnnotations(this);
	private final SessionEventSupport sessionEvents;
	
//	private int submittedOffset = 0;
	
	public TheoryEditor() {
		super();
		colorManager = new ColorManager();
		setSourceViewerConfiguration(new IsabelleTheoryConfiguration(this, colorManager));
//		setSourceViewerConfiguration(new TextSourceViewerConfiguration(EditorsUI.getPreferenceStore()));
		setDocumentProvider(new IsabelleFileDocumentProvider());
		
		Isabelle isabelle = IsabelleCorePlugin.getIsabelle();
		
		// create listeners for perspective change (scrolling & resize)
		this.viewerViewportListener = new IViewportListener() {
			
			@Override
			public void viewportChanged(int verticalOffset) {
				updateActivePerspective();
			}
		};
		
		this.viewerControlListener = new ControlAdapter() {

			@Override
			public void controlResized(ControlEvent e) {
				updateActivePerspective();
			}
		};
		
		isabelle.addSessionListener(sessionListener = new IIsabelleSessionListener() {
			
			@Override
			public void systemInit() {
				initIsabelleSystem();
			}

			@Override
			public void sessionInit(Session session) {
				initSession(session, getEditorInput());
			}
			
			@Override
			public void sessionShutdown(Session session) {
				shutdownSession(session, false);
			}
		});
		
		// When commands change (e.g. results from the prover), signal to update the document view,
		// e.g. new markups, etc.
		sessionEvents = new SessionEventSupport(EnumSet.of(SessionEventType.COMMAND)) {
			
			@Override
			protected SessionActor createSessionActor(Session session) {
				return new SafeSessionActor().commandsChanged(new ISessionCommandsListener() {
					
					@Override
					public void commandsChanged(Commands_Changed changed) {
						
						if (isabelleModel == null) {
							// no model available, so ignore
							return;
						}
						
						// avoid updating if commands are from a different document
						if (changed.nodes().contains(isabelleModel.name())) {
							refreshView();
						}
					}
				});
			}

			@Override
			protected void sessionInit(Session session) {
				// when the session is initialised refresh the view
				refreshView();
			}
		};
	}
	
	@Override
	public void init(IEditorSite site, IEditorInput input) throws PartInitException {
		
		// init Isabelle from core plugin
		Isabelle isabelle = IsabelleCorePlugin.getIsabelle();
		Session session = isabelle.getSession();
		if (session != null) {
			initSession(session, input);
		}
		
		super.init(site, input);
		
		activateContext();
	}
	
	/**
	 * Activate a context that this view uses. It will be tied to this editor
	 * activation events and will be removed when the editor is disposed.
	 */
	private void activateContext() {
		IContextService service = (IContextService) getSite().getService(IContextService.class);
		service.activateContext(EDITOR_SCOPE);
	}

	@Override
	public void createPartControl(Composite parent) {
		super.createPartControl(parent);
		
		// initialise perspective listeners after creating the control
		initPerspectiveListeners();
		updateActivePerspective();
	}

	private void initIsabelleSystem() {
		
		if (getEditorInput() != null) {
			// reload editor
			// TODO ask for save?
			reload();
		}
	}

	private void reload() {
		// reload input in the UI thread
		UIJob reloadJob = new UIJob("Reloading editor") {
			@Override
			public IStatus runInUIThread(IProgressMonitor monitor) {
				setInput(getEditorInput());
				return Status.OK_STATUS;
			}
		};
		reloadJob.schedule();
	}
	
	private void reloadOutline() {
		if (outlinePage == null) {
			return;
		}
		
		// reload outline in the UI thread
		UIJob reloadJob = new UIJob("Reloading outline") {
			@Override
			public IStatus runInUIThread(IProgressMonitor monitor) {
				outlinePage.reload();
				return Status.OK_STATUS;
			}
		};
		reloadJob.schedule();
	}
	
	/**
	 * Refreshes the text presentation
	 */
	private void refreshView() {
		final ISourceViewer viewer = getSourceViewer();
		if (viewer != null && viewer.getTextWidget() != null) {
			final Control control = viewer.getTextWidget();
			control.getDisplay().asyncExec(new Runnable() {
				
				@Override
				public void run() {
					if (!control.isDisposed()) {
						viewer.invalidateTextPresentation();
					}
				}
			});
		}
	}

	@Override
	public void dispose() {
		
		if (isabelleSession != null) {
			shutdownSession(isabelleSession, true);
		}
		
		disposePerspectiveListeners();
		
		// TODO review what happens if a second editor is opened for the same input
		Isabelle isabelle = IsabelleCorePlugin.getIsabelle();
		isabelle.removeSessionListener(sessionListener);
		
		markers.dispose();
		sessionEvents.dispose();
		
		colorManager.dispose();
		super.dispose();
	}
	
	private void initSession(Session session, IEditorInput input) {
		
		Assert.isTrue(this.isabelleSession == null);
		
		this.isabelleSession = session;
		initIsabelleModel();
		
		reloadOutline();
	}

	private void shutdownSession(Session session, boolean dispose) {
		
		if (this.isabelleSession == null) {
			return;
		}
		
		Assert.isTrue(this.isabelleSession == session);
		
		disposeIsabelleModel();
		this.isabelleSession = null;
		
		if (!dispose) {
			// not disposing, just shutting down the session
			reloadOutline();
		}
	}
	
	@Override
	protected void doSetInput(IEditorInput input) throws CoreException {
		
		// disconnect old model
		disposeIsabelleModel();
		
		super.doSetInput(input);
		
		this.documentRef = createDocumentRef(input);
		
		// connect to the new document
		initIsabelleModel();
		
		markers.updateAllAnnotations();
		
		// TODO need to reload?
		reloadOutline();
	}
	
	private void initIsabelleModel() {
		IDocument document = getDocument();
		if (document != null && documentRef != null && isabelleSession != null) {
			isabelleModel = DocumentModel$.MODULE$.init(isabelleSession, document, documentRef.getRef());
			
			// update active perspective in the UI thread
			getSite().getWorkbenchWindow().getWorkbench().getDisplay()
					.asyncExec(new Runnable() {

						@Override
						public void run() {
							updateActivePerspective();
						}
					});
//			isabelleModel.setSubmitOffset(submittedOffset);
			
			loadTheoryImports();
		}
	}
	
	private void initPerspectiveListeners() {
		// listen to scroll and resize events
		getSourceViewer().addViewportListener(viewerViewportListener);
		getSourceViewer().getTextWidget().addControlListener(viewerControlListener);
		
		// listen to document change as well
		this.documentListener = new DocumentListenerSupport(getSourceViewer(),
				new IDocumentListener() {

					@Override
					public void documentChanged(DocumentEvent event) {
						updateActivePerspective();
					}

					@Override
					public void documentAboutToBeChanged(DocumentEvent event) {
					}
				});
	}
	
	private void disposePerspectiveListeners() {
		ITextViewer viewer = getSourceViewer();
		if (viewer != null && viewer.getTextWidget() != null) {
			viewer.removeViewportListener(viewerViewportListener);
			viewer.getTextWidget().removeControlListener(viewerControlListener);
		}
		
		if (documentListener != null) {
			documentListener.dispose();
		}
	}
	
	/**
	 * Loads imported theories to Isabelle. Initialises document models for each of the dependency
	 * but does not open new editors.
	 * 
	 * TODO ask to open new editors as in jEdit?
	 */
	private void loadTheoryImports() {
		
		Isabelle isabelle = IsabelleCorePlugin.getIsabelle();
		Thy_Info theoryInfo = isabelle.getTheoryInfo();
		List<? extends Tuple2<DocumentRef, ?>> deps = TheoryInfoUtil.getDependencies(
				theoryInfo, Collections.singletonList(documentRef));
		
		// use set to avoid duplicates but preserve the order
		Set<DocumentRef> depRefs = new LinkedHashSet<DocumentRef>();
		for (Tuple2<DocumentRef, ?> dep : deps) {
			depRefs.add(dep._1());
		}
		
		// remove this editor
		depRefs.remove(documentRef);
		
		// check open editors which have been loaded
		for (IEditorPart editor : EditorUtil.getOpenEditors()) {
			if (editor instanceof TheoryEditor) {
				// TODO more generic instead of TheoryEditor, e.g. via IAdaptable?
				TheoryEditor theoryEditor = (TheoryEditor) editor;
//				if (theoryEditor.isabelleModel != null) {
//					// model is init
					depRefs.remove(theoryEditor.documentRef);
//				}
			}
		}
		
		if (depRefs.isEmpty()) {
			// no dependencies
			return;
		}
		
		IsabelleFileDocumentProvider docProvider = new IsabelleFileDocumentProvider();
		
		for (DocumentRef ref : depRefs) {

			// resolve document URI to load the file contents
			URI uri = URIThyLoad.resolveDocumentUri(ref.getRef());
			
			IFileStore fileStore;
			try {
				fileStore = EFS.getStore(uri);
			} catch (CoreException e) {
				// cannot load parent - skip
				IsabelleCorePlugin.log(e);
				continue;
			}
			
			// create the editor input to feed to document provider
			IEditorInput input = EditorUtil.getEditorInput(fileStore);
			
			try {
				docProvider.connect(input);
				IDocument document = docProvider.getDocument(input);
				// init document model
				DocumentModel model = DocumentModel$.MODULE$.init(isabelleSession, document, ref.getRef());
				// dispose immediately after initialisation
				model.dispose();
				docProvider.disconnect(input);
			} catch (CoreException e) {
				IsabelleUIPlugin.log(e.getMessage(), e);
			}
		}
	}
	
	private void disposeIsabelleModel() {
		if (isabelleModel != null) {
			isabelleModel.dispose();
		}
		isabelleModel = null;
	}

	public int getCaretPosition() {
		return getSourceViewer().getTextWidget().getCaretOffset();
	}
	
	public Snapshot getSnapshot() {
		
		if (isabelleModel == null) {
			return null;
		}
		
		return isabelleModel.snapshot();
	}
	
	/**
	 * Resolves a URI for the given editor input. For workspace files, the {@code platform:} URI
	 * scheme is used, for the rest it is resolved through IFileStore. Local filesystem files have
	 * {@code file:} URIs.
	 * 
	 * @param input
	 * @return the URI corresponding to the given editor input, or {@code null} if one cannot be
	 *         resolved.
	 * @throws URISyntaxException
	 */
	private static URI getInputURI(IEditorInput input) throws URISyntaxException {
		
		if (input == null) {
			return null;
		}
		
		// first check if the input is on a workspace file
		IFile file = ResourceUtil.getFile(input);
		if (file != null) {
			// encode the path as "platform:" URI
			return getResourceURI(file);
		}
		
		// otherwise try to resolve the URI or IPath
		ILocationProvider provider = (ILocationProvider) input.getAdapter(ILocationProvider.class);
		if (provider instanceof ILocationProviderExtension) {
			// get the URI, if available, and use it
			URI uri = ((ILocationProviderExtension) provider).getURI(input);
			if (uri != null) {
				return uri;
			}
		}
		
		if (provider != null) {
			IPath path = provider.getPath(input);
			if (path != null) {
				// the path can be either workspace, or file system
				// check if it is in the workspace first
				IResource resource = ResourcesPlugin.getWorkspace().getRoot().findMember(path);
				if (resource != null) {
					// found in workspace - use workspace path
					return getResourceURI(resource);
				} else {
					// absolute file system path
					return URIUtil.toURI(path);
				}
			}
		}
		
		return null;
	}
	
	/**
	 * Creates a workspace resource URI using {@code platform:} scheme.
	 * 
	 * @param resource
	 * @return
	 * @throws URISyntaxException
	 */
	private static URI getResourceURI(IResource resource) throws URISyntaxException {
		IPath path = resource.getFullPath();
		return URIThyLoad.createPlatformUri(path.toString());
	}
	
	private static DocumentRef createDocumentRef(IEditorInput input) {
		Assert.isNotNull(input);
		
		URI uri;
		try {
			uri = getInputURI(input);
		} catch (URISyntaxException e) {
			IsabelleUIPlugin.log(e.getMessage(), e);
			return null;
		}
		
		if (uri == null) {
			return null;
		}
		
//		String uriStr = uri.toString();
		String uriStr = URIPathEncoder.encodeAsPath(uri);
		
		Option<String> theoryNameOpt = Thy_Header.thy_name(uriStr);
		if (theoryNameOpt.isEmpty()) {
			IsabelleUIPlugin.log("Invalid theory name for URI: " + uriStr, null);
			return null;
		}
		
		String theoryName = theoryNameOpt.get();
		
		URI parentUri = getParentURI(uri);
//		String parentStr = parentUri.toString();
		String parentStr = URIPathEncoder.encodeAsPath(parentUri);
		
		return DocumentRef.create(uriStr, parentStr, theoryName);
	}
	
	/**
	 * Resolves parent URI for the given one. As proposed in
	 * http://stackoverflow.com/questions/10159186/how-to-get-parent-url-in-java
	 * 
	 * @param uri
	 * @return
	 */
	private static URI getParentURI(URI uri) {
		return uri.getPath().endsWith("/") ? uri.resolve("..") : uri.resolve(".");
	}
	
	public Session getIsabelleSession() {
		return isabelleSession;
	}
	
	public DocumentModel getIsabelleModel() {
		return isabelleModel;
	}
	
	public void submitToCaret() {
		
		if (isabelleModel == null) {
			// TODO show a message that model is not available?
//			System.out.println("Isabelle model is not available for " + getPathURI(getEditorInput()));
			return;
		}
		
		MessageDialog.openInformation(getSite().getShell(), "Functionality Disabled", 
				"Submitting to selection is currently disabled in favour of submitting active view.");
		// TODO remove altogether?
		
//		int caretOffset = getCaretPosition();
//		
//		IDocument document = getDocument();
//		
//		try {
//			
//			int caretLine = document.getLineOfOffset(caretOffset);
//			int caretLineEnd = getLineEndOffset(document, caretLine);
//			
//			System.out.println("Go to Command: " + getTitle() + ":[" + caretLine + "," + caretLineEnd + "]");
//			
//			isabelleModel.setSubmitOffset(caretLineEnd);
//			isabelleModel.updatePerspective();
//			
//			this.submittedOffset = caretLineEnd;
//			
//			// TODO review
//			reloadOutline();
//			
//		} catch (BadLocationException ex) {
//			IsabelleEclipsePlugin.log("Bad location in the document", ex);
//		}
	}
	
//	private static int getLineEndOffset(IDocument document, int line) throws BadLocationException {
//		return document.getLineOffset(line) + document.getLineLength(line);
//	}
	
	/**
	 * Updated the active perspective in the model. Finds the region currently
	 * visible in the editor and marks that in the model as its perspective -
	 * the area that should be submitted to the prover.
	 */
	private void updateActivePerspective() {
		
		if (getSourceViewer() == null) {
			return;
		}
		
		if (isabelleModel == null) {
			return;
		}
		
		IDocument document = getDocument();
		if (document == null) {
			return;
		}
		
		ILineRange visibleLines = JFaceTextUtil.getVisibleModelLines(getSourceViewer());
		
		int start = 0;
		int end = 0;
		if (visibleLines.getNumberOfLines() > 0 && visibleLines.getStartLine() >= 0) {
			// something is visible
			try {
				start = document.getLineOffset(visibleLines.getStartLine());
				int endLine = visibleLines.getStartLine() + visibleLines.getNumberOfLines();
				if (endLine >= document.getNumberOfLines() - 1) {
					end = document.getLength();
				} else {
					end = document.getLineOffset(endLine) + document.getLineLength(endLine);
				}
			} catch (BadLocationException e) {
				IsabelleUIPlugin.log(e.getMessage(), e);
			}
		}
		
		if (end < start) {
			end = start;
		}
		
		isabelleModel.setActivePerspective(start, end - start);
	}
	
	public IDocument getDocument() {
		IDocumentProvider provider = getDocumentProvider();
		if (provider == null) {
			return null;
		}
		
		return provider.getDocument(getEditorInput());
	}

	@Override
	public Object getAdapter(@SuppressWarnings("rawtypes") Class adapter) {
		if (IContentOutlinePage.class.equals(adapter)) {
			if (outlinePage == null) {
				outlinePage = new TheoryOutlinePage(this);
			}
			return outlinePage;
		}
		
		// expose the Isabelle model via IAdaptable
		if (DocumentModel.class.equals(adapter)) {
			return isabelleModel;
		}
		
		return super.getAdapter(adapter);
	}
	
	/**
	 * Selects and reveals the range in the editor.
	 * 
	 * @param selectRange
	 *            range to be selected
	 * @param highlightRange
	 *            range to be highlighted (e.g. with range annotation)
	 */
	public void selectInEditor(IRegion selectRange, IRegion highlightRange) {
		
		setHighlightRange(highlightRange.getOffset(), highlightRange.getLength(), true);
		selectAndReveal(selectRange.getOffset(), selectRange.getLength(),
				highlightRange.getOffset(), highlightRange.getLength());
		
		// TODO select in outline as well?
	}
	
	/**
	 * Selects the command in the editor.
	 * 
	 * @param command
	 *            the command to be selected, must be in the snapshot
	 * @param regionInCommand
	 *            region within the command to be selected, or {@code null} if the whole command is
	 *            to be selected
	 */
	public void setSelection(Command command, IRegion regionInCommand) {
		if (isabelleModel == null || command == null) {
			return;
		}
		
		// find the command in the snapshot
		Snapshot snapshot = getSnapshot();
		Option<Object> startOpt = snapshot.node().command_start(command);
		if (startOpt.isEmpty()) {
			// no command in the snapshot
			return;
		}
		
		// get the full command range
		int start = (Integer) startOpt.get();
		int length = command.length();
		IRegion cmdRange = new Region(start, length);
		
		// if available, calculate the region in command
		IRegion selectRange = cmdRange;
		if (regionInCommand != null) {
			selectRange = new Region(start + command.decode(regionInCommand.getOffset()), 
									 regionInCommand.getLength());
		}
		
		// reveal & select
		selectInEditor(selectRange, cmdRange);
	}
	
}

File: isabelle.eclipse.ui/src/isabelle/eclipse/ui/editors/TheoryEditor.scala
package isabelle.eclipse.ui.editors

import java.net.{URI, URISyntaxException}

import scala.actors.Actor._
import scala.collection.JavaConverters._

import org.eclipse.core.filesystem.EFS
import org.eclipse.core.runtime.CoreException
import org.eclipse.jface.dialogs.MessageDialog
import org.eclipse.jface.resource.{JFaceResources, LocalResourceManager}
import org.eclipse.jface.text.{IDocument, IRegion, ITextViewerExtension2, Region}
import org.eclipse.jface.text.source.IAnnotationModel
import org.eclipse.jface.viewers.{ISelectionChangedListener, SelectionChangedEvent}
import org.eclipse.swt.widgets.Composite
import org.eclipse.ui.{IEditorInput, IEditorSite, PartInitException}
import org.eclipse.ui.contexts.IContextService
import org.eclipse.ui.editors.text.TextEditor
import org.eclipse.ui.views.contentoutline.IContentOutlinePage

import isabelle.{Command, Document, Session, Thy_Header, Thy_Info}
import isabelle.Text.Range
import isabelle.eclipse.core.IsabelleCore
import isabelle.eclipse.core.app.Isabelle
import isabelle.eclipse.core.resource.URIThyLoad._
import isabelle.eclipse.core.text.{DocumentModel, EditDocumentModel, IsabelleDocument, ReadOnlyDocumentModel}
import isabelle.eclipse.core.util.AdapterUtil.adapt
import isabelle.eclipse.core.util.LoggingActor
import isabelle.eclipse.ui.editors.EditorUtil2.preserveScroll
import isabelle.eclipse.ui.internal.IsabelleImages
import isabelle.eclipse.ui.internal.IsabelleUIPlugin.{error, log}
import isabelle.eclipse.ui.util.JobUtil.uiJob
import isabelle.eclipse.ui.util.ResourceUtil
import isabelle.eclipse.ui.util.SWTUtil.asyncUnlessDisposed
import isabelle.eclipse.ui.views.outline.TheoryOutlinePage


/** The editor for Isabelle theory files.
  * 
  * Editor is backed by a Isabelle DocumentModel, which ensures sychronisation with the prover.
  * When the prover is available, the Isabelle editor #state is created, which adds functionality
  * that requires the active prover.
  *  
  * @author Andrius Velykis
  */
object TheoryEditor {

  val EDITOR_ID = "isabelle.eclipse.ui.theoryEditor"
  val EDITOR_SCOPE = "isabelle.eclipse.ui.theoryEditorScope"

}

class TheoryEditor extends TextEditor {

  import TheoryEditor._

  /** Resource manager for shared system resources that need to be disposed (e.g. colours) */
  val resourceManager = new LocalResourceManager(JFaceResources.getResources)

  /** The Isabelle-state of the editor (available after Isabelle is launched) */
  private var state: Option[State] = None
  private var init = false

  /** A flag to indicate that Isabelle document needs initialisation (do symbols) */
  @volatile private var needDocumentInit = false
  
  private var editable = true

  val outlinePage = new TheoryOutlinePage(this, getSourceViewer)
  // listen to outline page selection changes and highlight the selection in editor
  outlinePage.addSelectionChangedListener(new ISelectionChangedListener {
    override def selectionChanged(event: SelectionChangedEvent) =
      outlinePage.selectedRegionInEditor foreach (r => selectInEditor(r, r))
  })


  private val systemListener = LoggingActor {
    loop {
      react {
        // upon system init, initialise editor document - the symbols have changed then
        case Isabelle.SystemInit => {
          needDocumentInit = true
          uiJob("Initialising Isabelle Document") { initDocument() }
        }
        // when session is init/shutdown, update the editor state accordingly
        case Isabelle.SessionInit(session) => uiJob("Initialising Isabelle Editor") { initState(session) }
        case Isabelle.SessionShutdown(_) => uiJob("Updating Isabelle Editor") { disposeState() }
        case _ =>
      }
    }
  }

  {
    setSourceViewerConfiguration(new IsabelleTheoryConfiguration(this, resourceManager))
    setDocumentProvider(new IsabelleFileDocumentProvider)
  }

  @throws(classOf[PartInitException])
  override def init(site: IEditorSite, input: IEditorInput) {

    super.init(site, input)
    activateContext()

    init = true;
  }

  /**
   * Activate a context that this view uses. It will be tied to this editor activation events
   * and will be removed when the editor is disposed.
   */
  private def activateContext() {
    val service = adapt(getSite.getService _)(classOf[IContextService])
    service.foreach(_.activateContext(EDITOR_SCOPE))
  }

  @throws(classOf[CoreException])
  override protected def doSetInput(input: IEditorInput) {

    // disconnect the old state
    disposeState()

    super.doSetInput(input)

    if (init) {
      // connect & init state with the current session and the new document
      IsabelleCore.isabelle.session.foreach(initState(_, input))
    }
  }

  override def createPartControl(parent: Composite) {
    super.createPartControl(parent)

    // add listener to the isabelle app to react to session init
    val isabelle = IsabelleCore.isabelle
    isabelle.systemEvents += systemListener

    // init state if session is already available
    isabelle.session.foreach(initState(_))
  }

  private def createDocumentName(input: IEditorInput): Option[Document.Node.Name] = {

    // resolve document URI from input and then get the document name from URI
    // (needs to conform to specific pattern, as specified in Thy_Header)
    inputURI(input) flatMap { uri =>
      {

        val name = Thy_Header.thy_name(uri.toString) map { URINodeName(uri, _) }
        if (name.isEmpty) {
          log(error(msg = Some("Cannot resolve theory name for URI: " + uri.toString)))
        }

        name map toDocumentNodeName
      }
    }
  }

  private def inputURI(input: IEditorInput): Option[URI] = {
    try {
      ResourceUtil.getInputURI(input)
    } catch {
      case e: URISyntaxException => {
        log(error(Some(e)))
        None
      }
    }
  }

  private def initState(session: Session, input: IEditorInput = getEditorInput) {

    initDocument()
    
    val name = createDocumentName(input)
    
    val loaded = name forall (n => session.thy_load.loaded_theories(n.theory))
    if (loaded) {
      makeReadOnly()
    }

    val docModel = name map { n =>
      if (loaded) new ReadOnlyDocumentModel(session, document, n)
      else new EditDocumentModel(session, document, n)
    }

    state = docModel map (new State(_))
    state.foreach(_.init())

    reloadOutline()
  }


  private def makeReadOnly() {
    // use loaded file icon for the editor
    setTitleImage(resourceManager.createImageWithDefault(IsabelleImages.ISABELLE_LOADED_FILE))
    // disable editing as well
    editable = false
    getSourceViewer.setEditable(isEditable())
  }
  
  override def isEditable() = editable && super.isEditable()


  def document: IDocument = getDocumentProvider.getDocument(getEditorInput)
  
  private[editors] def annotationModel: Option[IAnnotationModel] = 
    Option(getDocumentProvider) flatMap (p => Option(p.getAnnotationModel(getEditorInput)))

  def isabelleModel: Option[DocumentModel] = state.map(_.isabelleModel)


  /**
   * Initialise document symbols
   */
  private def initDocument() = if (needDocumentInit) {
    document match {
      case isaDoc: IsabelleDocument => preserveScroll(getSourceViewer) {
        isaDoc.init()
        // also reset undo manager, because if there are undo actions,
        // they mess up encoded/decoded appearance
        getSourceViewer.resetPlugins
      }
      case _ =>
    }
    needDocumentInit = false
  }


  private def reloadOutline() = outlinePage.reload()


  override def dispose() {

    IsabelleCore.isabelle.systemEvents -= systemListener

    // TODO review what happens if a second editor is opened for the same input
    disposeState()
    resourceManager.dispose()
    super.dispose()
  }

  private def disposeState() {
    state.foreach(_.dispose())
    state = None
  }


  override def getAdapter(adapter: Class[_]): AnyRef = {

    val value =
      if (classOf[IContentOutlinePage] == adapter) {
        Some(outlinePage)

      } else if (classOf[DocumentModel] == adapter) {
        // expose the Isabelle model via IAdaptable
        state.map(_.isabelleModel)

      } else {
        None
      }

    value.getOrElse(super.getAdapter(adapter))
  }


  def caretPosition: Int = getSourceViewer.getTextWidget.getCaretOffset

  def submitToCaret() = state foreach { _ =>
    MessageDialog.openInformation(getSite.getShell, "Functionality Disabled",
      "Submitting to selection is currently disabled in favour of submitting active view.")
  }

  /**
   * Selects and reveals the range in the editor.
   *
   * @param selectRange  range to be selected
   * @param highlightRange  range to be highlighted (e.g. with range annotation)
   */
  def selectInEditor(selectRange: IRegion, highlightRange: IRegion) {

    setHighlightRange(highlightRange.getOffset, highlightRange.getLength, true)
    selectAndReveal(selectRange.getOffset, selectRange.getLength,
      highlightRange.getOffset, highlightRange.getLength)

    // TODO select in outline as well?
  }

  /**
   * Selects the command in the editor.
   *
   * @param command  the command to be selected, must be in the snapshot
   * @param regionInCommand  region within the command to be selected, or `None` if the
   *                         whole command is to be selected
   */
  def setSelection(command: Command, regionInCommand: Option[IRegion]) {

    // find the command in the snapshot
    val commandStart = isabelleModel.flatMap(_.snapshot.node.command_start(command))

    // get the full command range
    commandStart.map(new Region(_, command.length)) foreach { cmdRange =>
      {
        val rangeInCommand = regionInCommand.map(range =>
          new Region(cmdRange.getOffset + range.getOffset, range.getLength))

        // reveal & select
        selectInEditor(rangeInCommand.getOrElse(cmdRange), cmdRange)
      }
    }
  }


  /* The prover-active state */

  /** The prover state of the editor (and associated components that require the prover to be running) */
  private class State(val isabelleModel: DocumentModel) extends DocumentPerspectiveTracker {

    // When commands change (e.g. results from the prover), signal to update the document view,
    // e.g. new markups, etc.
    val commandChange = new CommandChangeHelper(isabelleModel)(refreshViewUI)

    override protected def textViewer = getSourceViewer
    
    private def control = Option(getSourceViewer) flatMap (v => Option(v.getTextWidget))

    val markers = new TheoryViewerAnnotations(
      Some(isabelleModel.snapshot),
      isabelleModel.document,
      annotationModel,
      Option(EditorUtil.getResource(getEditorInput)))

    def init() {
      
      isabelleModel.init()
      commandChange.init()
      
      initPerspective()

      loadTheoryImports()
      
      // refresh all after initialisation
      refreshView()
    }

    def dispose() {
      commandChange.dispose()
      disposePerspective()
    }

    def refreshViewUI(changedRanges: Option[List[Range]] = None) =
      asyncUnlessDisposed(control)(refreshView(changedRanges))

    /**
     * Refreshes the text presentation.
     * 
     * Must be called in the UI thread.
     */
    def refreshView(changedRanges: Option[List[Range]] = None) {

      // regenerate annotations for the changed ranges
      markers.updateAnnotations(changedRanges)

      // After setting annotations, queue viewer refresh in the UI thread.
      // 
      // This is necessary because there seems to be a race condition with annotation model
      // when some refreshes are being lost, so doing it explicitly here.
      // AnnotationRepainter schedules refresh in UI thread using asyncExec, so add one more after
      asyncUnlessDisposed(control) {
        (Option(getSourceViewer), changedRanges) match {
          case (Some(textViewer: ITextViewerExtension2), Some(ranges)) =>
            ranges foreach (r => textViewer.invalidateTextPresentation(r.start, r.length))

          case (Some(viewer), _) => viewer.invalidateTextPresentation()

          case _ =>
        }
      }
    }

    /**
     * Loads imported theories to Isabelle. Initialises document models for each of the dependency
     * but does not open new editors.
     *
     * TODO ask to open new editors as in jEdit?
     */
    private def loadTheoryImports() {

      val nodes = pendingDependencies()

      nodes.foreach(node => {

        // resolve document URI to load the file contents
        val uri = resolveDocumentUri(node)

        val fileStore =
          try {
            Some(EFS.getStore(uri))
          } catch {
            case e: CoreException => {
              // cannot load parent - skip
              log(error(Some(e)))
              None
            }
          }

        // create the editor input to feed to document provider
        fileStore.map(EditorUtil.getEditorInput) foreach { input =>
          withDocument(input) { document =>
            {
              // init document model
              val model = new EditDocumentModel(isabelleModel.session, document, node)
              model.init()
              // dispose immediately after initialisation
              model.dispose()
            }
          }
        }

      })
    }

    private def pendingDependencies(): List[Document.Node.Name] = {

      val thyInfo = new Thy_Info(isabelleModel.session.thy_load)

      val currentName = isabelleModel.name

      // get the dependencies for this name and filter the duplicates as well as this editor
      val dependencies = thyInfo.dependencies(true, List(currentName)).deps
      val dependencyNodes = dependencies.map(_.name).distinct.filter(_ != currentName)
      
      // get document models for each open editor and resolve their names
      val loadedNodes = EditorUtil.getOpenEditors.asScala.map(
        editor => adapt(editor.getAdapter _)(classOf[DocumentModel])).flatten.map(_.name).toSet

      dependencyNodes.filterNot(loadedNodes.contains)
    }

    private def withDocument(input: AnyRef)(f: IDocument => Unit) =
      try {
        val provider = getDocumentProvider

        provider.connect(input)
        val document = provider.getDocument(input)
        f(document)
        provider.disconnect(input)

      } catch {
        case e: CoreException => log(error(Some(e)))
      }
  }

}
